### PR TITLE
Brokers must validate parameters passed to provision, update and bind

### PR DESCRIPTION
**Pull request by mattmcneeney**
With work underway regarding service brokers declaring a supported schema for the parameters field ( #59 ), I think it makes sense that brokers ensure that any parameters provided via the provision, update and bind calls are valid.

### CODE REVIEW

**Comment by mattmcneeney** [comment]
I've fixed my membership regarding the cfdreddbot message. Please can an owner close and reopen this PR, thanks!

File: spec.md [hunk 1]
```diff
@@ -570,7 +570,7 @@ the resource it creates.
 | plan_id*  | string  | ID of the plan from the catalog.  |
 | app_guid  | string  | Deprecated in favor of <code>bind\_resource.app\_guid</code>. GUID of an application associated with the binding to be created.  |
 | bind_resource  | JSON object  | A JSON object that contains data for platform resources associated with the binding to be created. Current valid values include <code>app\_guid</code> for [credentials](#types-of-binding) and <code>route</code> for [route services](#route_services).  |
-| parameters | JSON object  |  Configuration options for the service binding. An opaque object, controller treats this as a blob.  |   |
+| parameters | JSON object  |  Configuration options for the service binding. An opaque object, controller treats this as a blob. Brokers must validate these parameters to ensure that they are valid. |   |
```
Comment (avade) [inline spec.md]: "Brokers should ensure that the client has provided valid configuration parameters and values for the operation"

Comment (avade) [inline spec.md]: I think this is clearer advice.

Comment (mattmcneeney) [inline spec.md]: I agree, thanks. I've updated the branch!

**Comment by avade** [comment]
LGTM

**Comment by angarg12** [comment]
LGTM

**Comment by duglin** [comment]
Even though we're not using RFC2119 keywords (yet), I'm glad you used "should" instead of "must". I was a little uncomfortable with the idea that an implementation couldn't have really smart error/typo recovery if they wanted.

So, what should they do if they do find something wrong? Do we want to RECOMMEND that they generate an error and reject the request? W/o guidance someone could interpret it as valid to just ignore the bad parameters and keep processing.  I'd suggest we generate an error and reject the request.

Also, should this be moved to a more global spot in the spec and apply to all fields?

**Comment by vaikas-google** [comment]
One would expect that any parameters (not 'parameters' but any input) passed in should be validated by the broker and handled appropriately? Isn't the point we're trying to make here that even though there's a schema that you can specify for the parameters you expect, ultimately the broker is responsible for validating the configuration parameters because to the controller, it's opaque. Even if the controller could validate it, a broker should always validate its inputs.
I think we might want to give guidance on what the error message returned should look like, for example, we might look at the k8s api conventions and use that as a guidance on how to help the developers to convey meaningful error messages back to the controller.
https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#validation

**Comment by mattmcneeney** [comment]
I agree that we should give guidance on returning meaningful error messages across the board. Should we handle that in a separate issue since that applies to the whole spec rather than just input parameters?

**Comment by mattmcneeney** [comment]
Just raising attention to this PR again. Discussion above suggests we need more guidance on returning meaningful error message across the API, but is this a useful standalone addition for now?

**Comment by avade** [comment]
> Discussion above suggests we need more guidance on returning meaningful error message across the API, but is this a useful standalone addition for now?

Yes, I think that adds value as a standalone addition. The input in the config parameters is not crafted by the platform, so I feel more prone to error.

Can you change `should` to `SHOULD` to meet RFC2119 keywords spec.

**Comment by mattmcneeney** [comment]
All `SHOULD`s are now uppercase.

**Comment by angarg12** [comment]
Is this ready for another round of lgtm?

**Comment by mattmcneeney** [comment]
I hope so!

**Comment by duglin** [comment]
LGTM

**Comment by mattmcneeney** [comment]
Please can we get two more on this (mainly so I don't have to fix more merge conflicts!) 😄

**Comment by avade** [comment]
LGTM

**Comment by avade** [comment]
@fmui, @pmorie, @shalako, @vaikas - needs one more LGTM.

**Comment by vaikas-google** [comment]
LGTM

**Pull request merged**